### PR TITLE
advanced tutorial

### PR DESCRIPTION
**Pull request by L1aoXingyu**

### CODE REVIEW

File: docs/source/tutorials/Advanced_Tutorial.md [hunk 1]
```diff
@@ -0,0 +1,240 @@
...
+
+>> rank: 0, tensor shape: oneflow.Size([32, 16384])
+>> rank: 1, tensor shape: oneflow.Size([32, 16384])
+```
+
+Through the result, you can find that `y` has been split along with `axis=1` on 2 GPUs.
+
+### Large MLP models
+
+MLP is very popular in transformer-based models. Assume we have a huge MLP model and its very large hidden size makes it difficult to fit into a single GPU.
```
Comment (Ldpe2G) [inline docs/source/tutorials/Advanced_Tutorial.md:54]:
```suggestion
MLP is very popular in transformer-based models. Assume we have a huge MLP model with huge hidden size which makes it difficult to fit into a single GPU.
```

File: docs/source/tutorials/Advanced_Tutorial.md [hunk 2]
```diff
@@ -0,0 +1,240 @@
...
+
+We can then split the model weights across GPUs in a 2D mesh while you still write your model in a familiar way.
```
Comment (Ldpe2G) [inline docs/source/tutorials/Advanced_Tutorial.md:56]: 下面的历史 模型的 sbp 是 [broadcast, split] ，感觉这里描述的权值 切割成 2d 网格形式有点不符？

File: docs/source/tutorials/Advanced_Tutorial.md [hunk 3]
```diff
@@ -0,0 +1,240 @@
...
+)
+
+# set a Linear placement to last stage
+# set it manually 
+self.head = Linear(embed_dim, num_classes, layer_idx=-1).to_global(placement=dist.get_layer_placement(-1))
+# use `layer_idx` API
+self.head = Linear(embed_dim, num_classes, layer_idx=-1)
+```
+
+After configuring models placement, you need to add the input placement transition across different stages. In LiBai, we set a `layer_idx` attribute in each `nn.Module`, so you can simply add `to_global` in `forward` to implement input placement transition.
```
Comment (Ldpe2G) [inline docs/source/tutorials/Advanced_Tutorial.md:145]:
```suggestion
After configuring model‘s placement, you need to add the input placement transition across different stages. In LiBai, we set a `layer_idx` attribute in each `nn.Module`, so you can simply add `to_global` in `forward` to implement input placement transition.
```

File: docs/source/tutorials/Advanced_Tutorial.md [hunk 4]
```diff
@@ -0,0 +1,240 @@
+# Advanced Tutorial
+
+## Define your own parallel model with LiBai.layers
+
+### Large-scale FC
+
+Let's say that you have a huge fully-connected-layer for large-scale classification (e.g., 1000w classes), which makes it impossible to fit into a single GPU.
+
+Don't worry, with help of `LiBai.layers`, you can write models in a familiar way that you used to write models for a single GPU. We give a simple example showing how to write a tensor-parallel fully-connected-layer with 2 GPUs.
```
Comment (rentainhe) [inline docs/source/tutorials/Advanced_Tutorial.md:9]: `with help of` -> `with the help of`，感觉这样比较合适

File: docs/source/tutorials/Advanced_Tutorial.md [hunk 5]
```diff
@@ -0,0 +1,240 @@
+# Advanced Tutorial
```
Comment (rentainhe) [inline docs/source/tutorials/Advanced_Tutorial.md:1]: 感觉可以单独开一个advanced_tutorials文件夹，然后把这个教程放进去，可能后续还要加一些别的advanced教程

File: docs/source/tutorials/Advanced_Tutorial.md [hunk 6]
```diff
@@ -0,0 +1,240 @@
...
+
+## Define your own parallel model with LiBai.layers
+
+### Large-scale FC
+
+Let's say that you have a huge fully-connected-layer for large-scale classification (e.g., 1000w classes), which makes it impossible to fit into a single GPU.
```
Comment (CPFLAME) [inline docs/source/tutorials/Advanced_Tutorial.md:7]:
```suggestion
Suppose you have a huge fully-connected-layer for large-scale classification (e.g., 1000w classes), which makes it impossible to fit into a single GPU.
```

File: docs/source/tutorials/Advanced_Tutorial.md [hunk 7]
```diff
@@ -0,0 +1,240 @@
...
+```shell
+python3 -m oneflow.distributed.launch --nproc_per_node 4 huge_mlp_example.py
+
+>> rank: 2, tensor shape: oneflow.Size([16, 1024])
+>> rank: 3, tensor shape: oneflow.Size([16, 1024])
+>> rank: 1, tensor shape: oneflow.Size([16, 1024])
+>> rank: 0, tensor shape: oneflow.Size([16, 1024])
+```
+
+From above, you can see that data are split into 2 groups for data parallel and weights are split into 2 groups for model parallel. So this simple example just implements a 2D parallel.
```
Comment (CPFLAME) [inline docs/source/tutorials/Advanced_Tutorial.md:111]: 上文用的 `tensor parallel` 这里用的`model  parallel`, 最好统一一下

Comment (rentainhe) [reply docs/source/tutorials/Advanced_Tutorial.md]: 是不是要有什么地方再强调一下模型并行是tensor parallel, 我发现官网叫model parallel, 但是我们内部叫tensor parallel 😆

File: docs/source/tutorials/advanced_tutorials/customize_parallel.md [hunk 8]
```diff
@@ -0,0 +1,240 @@
...
+
+>> rank: 0, tensor shape: oneflow.Size([32, 16384])
+>> rank: 1, tensor shape: oneflow.Size([32, 16384])
+```
+
+Through the result, you can find that `y` has been split along with `axis=1` on 2 GPUs.
+
+### Large MLP models
+
+MLP is very popular in transformer-based models. Assume we have a huge MLP model with huge hidden size which makes it difficult to fit into a single GPU.
```
Comment (thinksoso) [inline docs/source/tutorials/advanced_tutorials/customize_parallel.md:54]: 上一节是Suppose you，这一节变成了Assume we, 是不是表达一致更通顺一些~

File: docs/source/tutorials/advanced_tutorials/customize_parallel.md [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+        ...
+        self.layer_idx = layer_idx
+        ...
+
+    def forward(self, hidden_states):
+        hidden_states = hidden_states.to_global(placement=dist.get_layer_placement(self.layer_idx))
+        ...
+```
+
+After configuring models and data placement, the only thing that needs to do is set the distributed configuration.
```
Comment (thinksoso) [inline docs/source/tutorials/advanced_tutorials/customize_parallel.md:159]: set -> setting

File: docs/source/tutorials/advanced_tutorials/customize_parallel.md [hunk 10]
```diff
@@ -0,0 +1,240 @@
...
+# set pipeline stages to 2
+train.dist.pipeline_parallel_size = 2
+
+# set model layers for pipeline
+train.dist.pipeline_num_layers = hidden_layers
+```
+
+### Introduction of 1F1B pipeline parallel
+
+First of all, we will introduce you GPipe for your better understanding. In GPipe, when the forward passes of all microbatches finish, the backward passes would be executed (shown in below).
```
Comment (thinksoso) [inline docs/source/tutorials/advanced_tutorials/customize_parallel.md:171]: introduce sb. to sth. 或者 introduce sth. to sb. 好像没有不加to的情况

File: docs/source/tutorials/Advanced_Tutorial.md (hunk 7 above)
Comment (L1aoXingyu) [reply docs/source/tutorials/Advanced_Tutorial.md]: 我写成 tensor model parallel 吧

Comment (thinksoso) [reply docs/source/tutorials/Advanced_Tutorial.md]: 这里的2D并行看起来不像是colossal ai的2d tensor parallel，求问为什么叫2D

Comment (L1aoXingyu) [reply docs/source/tutorials/Advanced_Tutorial.md]: 数据并行+模型并行，所以叫 2d 并行

**Review by rentainhe — APPROVED** [review]

**Review by Ldpe2G — APPROVED** [review]

**Review by CPFLAME — APPROVED** [review]

**Review by thinksoso — APPROVED** [review]
